models/staging/ga4/stg_ga4__sessions_traffic_sources.sql

```diff
@@ -21,24 +23,14 @@ set_default_channel_grouping as (
 session_source as (
     select    
         session_key,
-
-        (case
-          when FIRST_VALUE(source) OVER (PARTITION BY session_key ORDER BY event_timestamp ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) is null then "(direct)"
-          else FIRST_VALUE(source) OVER (PARTITION BY session_key ORDER BY event_timestamp ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) end
-        ) as session_source,
-
-        (case 
-          when FIRST_VALUE(medium) OVER (PARTITION BY session_key ORDER BY event_timestamp ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) is null then "(none)"
-          else FIRST_VALUE(medium) OVER (PARTITION BY session_key ORDER BY event_timestamp ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) end
-        ) as session_medium,
-
-        (case 
-          when FIRST_VALUE(campaign) OVER (PARTITION BY session_key ORDER BY event_timestamp ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) is null then "(direct)"
-          else FIRST_VALUE(campaign) OVER (PARTITION BY session_key ORDER BY event_timestamp ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) end
-        ) as session_campaign,
-
-        FIRST_VALUE(default_channel_grouping) OVER (PARTITION BY session_key ORDER BY event_timestamp ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS session_default_channel_grouping
+        COALESCE(FIRST_VALUE(source) OVER (session_window), "(direct)") AS session_source,
+        COALESCE(FIRST_VALUE(medium) OVER (session_window), "(none)") AS session_medium,
+        COALESCE(FIRST_VALUE(campaign) OVER (session_window), "(direct)") AS session_campaign,
```

Comment:
I realize this issue existed before this PR, but isn't it problematic to assume that a null `campaign` value is `(direct)`? There is plenty of non-direct traffic that never has a campaign assigned. 

Seems like `(none)` would be more appropriate for an empty campaign value, but I might be missing something.

Reply:
Honestly I'm not too sure why we put values in here rather than leaving them NULL and letting the end user's viz tool take care of it, but I think (?) that the (none) and (direct) values are attempting to emulate what GA4's UI shows.
![Screen Shot 2022-10-31 at 13 18 59](https://user-images.githubusercontent.com/12062/198909131-9ec09023-79eb-45fc-af6b-5fabc9021043.png)
![Screen Shot 2022-10-31 at 13 19 38](https://user-images.githubusercontent.com/12062/198909156-e88a8f45-7978-4a52-a34f-0d155cbd11da.png)

In that second screenshot we can see that there are separate (direct) and (not set) values, so looks like GA4 is being a bit more conditional about what it says - I assume the default value it puts in the campaign field depends on the source and/or medium.

More general question then, do we want to be trying to emulate that stuff, or should we leave it all NULL and let the end user figure out what they want to say?

Reply:
(referral) and (organic) seem a little different to me because they are actually based on some positive evidence - not just the absence of any other value. (referral) happens when there's an external, non-excluded referrer; organic is IIRC used with Google search referrals specifically.

I don't think they fill in anything at all for sessions that have no clear origin?

I guess I'd like it if the materialized tables (such as dim_ga4__sessions which uses this view) did not do any of this null-filling kind of thing, but we could do it in mart views. Then it's easy to keep composing more views on top of the materialized tables (for example, I'm probably about to have to do some extra work to find the first and last non-none session attribution per user, which would be an easy `FIRST/LAST_VALUE(.. IGNORE NULLS)` if null).

If we are keeping it, I don't have strong feelings about what the value should be, but the other fields mostly use (none), so would that be better than (not set)?

Reply:
You know, we're getting a bit off-topic here. I've pushed (none) to align with the others, how do you feel about that?